design-proposals/003-credentials.md

```diff
@@ -0,0 +1,158 @@
+
+### Credential backend model
+
+Credential backend model is key-value pairs grouped together under the same identifier. The identifier is identical in value to the identifier used in the view model. Keys of the pairs correspond to the property names needed for the Spring Boot configuration of the Camel components.
+
+## REST API
+
+To interact with the credentials API:
+
+| HTTP Verb | Path | Description |
+| --------- | ---- | ----------- |
+| GET | /api/{version}/credentials/{connector} | List all the ways a credential can be acquired for a {connector} |
+| POST | /api/{version}/credentials/{connector} | Acquire new credential for connector with label {connector} |
+| \* | /api/{version}/credentials/{connector}/acquire/\*\* | Handle interaction with 3rd party services during credential acquisition |
+| GET | /api/{version}/credentials | List all credentials created by the current tennant |
+| DELETE | /api/{version}/credentials/{connector}/{id} | Remove credential from the system |
+
+### Examples
+
+#### Acquisition example
+
+Here is an example of credential workflow starting with acquisition for Salesforce:
+ 1. UI lists all methods of credential acquisition
+```http
+GET /api/v1/credentials/salesforce HTTP/1.1
+Accept: application/json
+```
+```json
+{
+  "response": {
+    "methods": [{
+      "type": "jwt-oauth",
+      "label": "Connect to Salesforce",
+      "icon": "x-salesforce",
+      "description": "You will be redirected to Salesforce to authorize OAuth usage."
+    }]
+  }
+}
+```
+ 2. Citizen user chooses "Connect to Salesforce", this results in creation of new credential acquisition:
+```http
+POST /api/v1/credentials/salesforce HTTP/1.1
+Content-Type: application/json
+Accept: application/json
+```
+```json
+{
+  "input": {
+    "method": "jwt-oauth",
+    "returnUrl": "https://{syndesis-ui}/..."
+  }
+}
+```
+```json
+{
+  "response": {
+    "action": "redirect",
+    "url": "https://login.salesforce.com/services/oauth2/authorize?..."
+  }
+}
+```
+3. UI redirects user to the specified `url`, and Salesforce in turn redirects to OAuth callback:
+```http
+GET /services/oauth2/authorize?... HTTP/1.1
+Host: login.salesforce.com
+```
+```http
+HTTP/1.1 302 Moved Temporarily
+Location: https://{{syndesis-rest}/api/v1/credentials/salesforce/acquire?code=...
+```
+4. Acquisition handler for Salesforce handles the request, extracts the `code` and returns to the URL provided by the UI in 2. (`returnUrl`)
+```http
+GET /api/v1/credentials/salesforce/acquire?code=... HTTP/1.1
+```
+```http
+HTTP/1.1 302 Moved Temporarily
+Location: https://{syndesis-ui}/...
+```
+5. Acquisition handler for Salesforce issues an out of bounds request to Salesforce that replies with the OAuth tokens
+```http
+POST /services/oauth2/token HTTP/1.1
+Host: login.salesforce.com
+```
+```http
+grant_type=authorization_code&client_id={sindesis OAuth client id}&client_secret={syndesisOauth client secret}&redirect_uri={syndesis ui returnUrl}&code=...
```

Comment:
May need to embed state too which is an optional field for the oauth flow. Not sure what @gashcrumb thinks about that?